src/lib/rule2E.ts

```diff
@@ -0,0 +1,85 @@
+import {Context, getDefaultContext} from './context';
+import {computeTextAlternative} from './compute_text_alternative';
+
+// Input types that imply role 'textbox'
+const TEXTBOX_INPUT_TYPES = ['email', 'tel', 'text', 'url'];
+
+/**
+ * Checks if input elem has role 'textbox' or has a list attribute
+ * @param elem - input element whose role is being calculated
+ * @return - true if elem has role 'textbox' or list attribute,
+ * false otherwise.
+ */
+function isTextboxOrListInput(elem: HTMLInputElement): boolean {
+  const inputType = elem.getAttribute('type')?.toLowerCase() ?? '';
+  return TEXTBOX_INPUT_TYPES.includes(inputType) || elem.hasAttribute('list');
```

Comment:
I think that the explicit roles should be handled by rule 2F, because for 2E to accept a node it already has to be part of a name, which means that 2F will calculate text alternatives for the entire subtree (which should include any text content from elements with role=textbox). Does that sound right to you? I could add some tests to compute_text_alternative_tests.ts to verify this.

A native textbox alternative becomes a combobox when a list attribute is present, and comboboxes happen to be handled in the same way as textboxes, so that's why I grouped these two conditions together: they both apply to `<input>`s and they both handle them in the same way (according the examples I could find online), namely by returning the `.value` of the `<input>`.

I can look into a more readable design or better comments if you think that would help here?

Reply:
`<input type="number" list="foo">` is neither a `textbox` nor a `combobox` but it will return true here.

If you rename the function to `isTextboxOrCombobox` and remove the ` || elem.hasAttribute('list')` does that achieve what you want?

Reply:
Ah yes, of course, thanks! And yep what you've described there will do the job. In my latest push I have actually already renamed the function but I'll remove the `elem.hasAttribute('list')` in the next one.